### DISCUSSION TITLE
Streamlines evenly sampled in time

### DISCUSSION THREAD

**Discussion by bheiles** [pyvista/pyvista#1966] [Q&A] [answered]
Hi! First of all thanks **a lot** for pyvista, i have been using for a month and find it quite easy and well documented, and of course quite powerful. I have a question regarding the generation of streamlines in 3D. I am trying to use the streamlines_from_source filter to generate snapshots of particles moving in a simulation at regular times. Only problem is, i cant seem to do that using either RK integration schemes. I tried tricking it by asking for a very small time step but it doesnt work. My question is, is there any chance i can manage to do that using that filter? Would you have advice in case I need to implement it myself?

Thanks!

**Comment by MatthewFlamm** [comment, marked as answer]
> I am trying to use the streamlines_from_source filter to generate snapshots of particles moving in a simulation at regular times. 

Do you have a time invariant or time varying velocity field?  For the former, the integrators work in space units rather than time units.  But you could interpolate the resulting lines to get equispaced time points along the line.  This will result in a loss of precision however.  

For the latter, this is going to be difficult to implement in pyvista since time varying datasets are not well (or at all?) supported.  There is a vtk class for this though:

https://vtk.org/doc/nightly/html/classvtkTemporalStreamTracer.html

**Reply by bheiles** [reply]
I have a time invariant field. What you're suggesting would work well if the time of integration was somewhat linear/regular. However when i look at it, follows a somewhat exponentiel curve... The interpolation is what i have been trying to do but it really does not work well. Any other suggestion?

**Reply by MatthewFlamm** [reply]
> However when i look at it, follows a somewhat exponentiel curve.

I think it will be hard to get semi equally spaced points in time unless the flow and/or mesh are very regular.  But if you have enough points along the line, you should still get reasonable accuracy by interpolating.  This can be done by lowering `max_step_length` and `max_error`

**Reply by bheiles** [reply]
Ok i see what you mean, that is also what i tried doing. I have two issues : first, what are the units of step lengths and of integration time? Second, if i do it this way, that means that i need to scale down all the step lengths for all streamlines based on the maximum speed in the geometry. Thus i will generate an immense amount of data for all streamlines just to abide by a min sampling rate in time for the fastest streamlines.

**Reply by MatthewFlamm** [reply]
Yes.  I still think you should be able to use the original data with the exponential-like distribution of time steps.  Since the simulation results vary in space on the length scale of a single cell of the mesh, as long as your steps are smaller than a mesh (this is easy to control), you should have multiple streamline points inside most intersecting cells.

**Reply by MatthewFlamm** [reply]
> first, what are the units of step lengths and of integration time?

The units of step lengths are either in cell lengths or in physical units.  There is no concept of time that is exposed to the end user through vtk or pyvista.

**Reply by bheiles** [reply]
Ok thanks ill try this ! And the unit of integration time?

**Reply by MatthewFlamm** [reply]
> Ok thanks ill try this ! And the unit of integration time?

It would be in whatever units your velocity field is in.  For example, velocity of m/s would have integration time in seconds.  velocity in km/hr would have time units of hr.

**Reply by bheiles** [reply]
Ok super thanks Matthew!

**Reply by bheiles** [reply]
Hi @MatthewFlamm ! So I've been trying to experiment with what you suggested... I have velocity in mm/s and my max velocity normed is around 0.100mm/s. I have tried a wide range of step parameters but I can not reach a lower integration time than around 300,000. This is is reached by taking step length extremely low as: initial=1e-10, min=1e-11, max=2e-7 and a max error at 1e-15. I am pretty sure there is something Im not doing right... Do you have any idea what that would be?

**Comment by MatthewFlamm** [comment]
I think you are right that the integration time is not in time units but in the arc length unit of the integration.  I should be able to cook up a simple example to figure out the actual physical time, and we could potentially add this to pyvista implementation under the hood as an option.

**Comment by MatthewFlamm** [comment]
This example still makes me believe that it is in real physical time units despite the confusion that the vtk documentation creates.

Here the cell length is chosen to be 2 (diagonally), the domain is 10 length units long, and velocity of 1 length unit/time unit in the direction of interest.  I chose RK2 integrator as it has a constant step size.  The step size is 0.1.

We should expect that using cell length units, we should take around 50 steps. Since the cell size is 2, the physical step size is 0.2.  When using physical units, we should take 100 steps.  The last time point should be close to 10.0

```py
import pyvista
import numpy as np
grid = pyvista.UniformGrid((11, 11, 11), (1, 1, np.sqrt(2)))
velocity = np.zeros((grid.n_points, 3))
velocity[:, 0] = 1.0
grid["velocity"] = velocity

stream = grid.streamlines(
    start_position=(0, 5, 5), 
    vectors="velocity",
    integrator_type=2, 
    initial_step_length=0.1, 
    integration_direction="forward", 
    step_unit="cl"
)
print(stream["IntegrationTime"][-5:])
print(stream["IntegrationTime"].shape)

stream = grid.streamlines(
    start_position=(0, 5, 5), 
    vectors="velocity",
    integrator_type=2, 
    initial_step_length=0.1, 
    integration_direction="forward", 
    step_unit="l"
)
print(stream["IntegrationTime"][-5:])
print(stream["IntegrationTime"].shape)
```




```txt
[ 9.2  9.4  9.6  9.8 10. ]
(51,)
[ 9.6  9.7  9.8  9.9 10. ]
(101,)
```

Running with `velocity[:, 0] = 2.0` should yield the same number of steps but an ending time of around 5.0.

```txt
[4.6 4.7 4.8 4.9 5. ]
(51,)
[4.8  4.85 4.9  4.95 5.  ]
(101,)
```

**Reply by bheiles** [reply]
1. Ok so I tried your example and it works the same for me. I have noticed that when you don't set the start position, you get less points in the end (max of integration_time is 5.2), even if you enforce a max_time condition.
2. I have the feeling something might be wrong with the mesh I am using then. I have computed cell sizes for my dataset and here is what I see:


![image](https://user-images.githubusercontent.com/32468395/148045071-907f10bf-c531-4406-ad21-313931390a2d.png)

In your example I see this:
![image](https://user-images.githubusercontent.com/32468395/148045157-55168a34-1316-43a0-8794-25cd04e7d908.png)


Maybe the fact that my bounds don't start at 0 are a problem? Thanks so much for your help

**Reply by MatthewFlamm** [reply]
When you don't set a start_position, a sphere source in the center of the mesh is used for seed points.  So the integration time would be less in this example, so that makes sense to me.

There are multiple possibilities here.

1. You have a non open system, or separated flow and some streamlines never terminate.
2. Some streamlines enter cells with very low velocity and thus take very long steps in physical time.  Since vtk doesn't get information about the numerical scheme you are using in CFD, numerical issues can cause lines to get too close to walls in my experience.

**Reply by bheiles** [reply]
Ok so in configuration 2 where the integration time is way above the time step fixed in the streamlines filter, I guess assuming constant velocity in between large integration time steps would be the best. That means that to retrieve even time sampling I would then fix an initial_step_length with RK2 for all streamlines I am trying to generate and if the integration time step output is higher than the initial_step_length I will just assume it's the same velocity and calculate it's new position manually. Cumbersome but it should work right?

**Reply by MatthewFlamm** [reply]
It seems like you are still trying to get pyvista/vtk to output equally spaced time values, and I don't think that is possible since the integration is done in length units as the independent variable.  I think the only way to align the time values is to interpolate.

This example interpolates onto the minimum total intregation time.  You can build in logic to capture the data for streamlines that have longer integration times too.

```py
import pyvista
import numpy as np
grid = pyvista.UniformGrid((11, 11, 11), (1, 1, np.sqrt(2)))
velocity = np.zeros((grid.n_points, 3))
velocity[:, 0] = grid.points[:, 1]
grid["velocity"] = velocity

stream = grid.streamlines(
    pointa=(0, 4, 5),
    pointb=(0, 6, 5),
    n_points=3, 
    vectors="velocity",
    integrator_type=2, 
    initial_step_length=0.1, 
    integration_direction="forward", 
    step_unit="l"
)
individual_streams = stream.split_bodies()
for s in individual_streams:
    print(s["IntegrationTime"][-5:])
min_time = np.min([s["IntegrationTime"][-1] for s in individual_streams])

def interpolate_time(mesh, times):
            
    xprobe = np.interp(times, mesh["IntegrationTime"], mesh.points[:, 0])
    yprobe = np.interp(times, mesh["IntegrationTime"], mesh.points[:, 1])
    zprobe = np.interp(times, mesh["IntegrationTime"], mesh.points[:, 2])
    return pyvista.PolyData(np.vstack((xprobe, yprobe, zprobe)).T).sample(mesh)

interpolated_meshes = [interpolate_time(s, np.linspace(0, min_time, 4)) for s in individual_streams]
for m in interpolated_meshes:
    print(m["IntegrationTime"])
```

gives

```txt
[1.9   1.925 1.95  1.975 2.   ]
[1.62857148 1.65000006 1.67142863 1.6928572  1.71428577]
[1.42499996 1.44374996 1.46249996 1.48124996 1.49999996]
[1.26666667 1.28333333 1.3        1.31666667 1.33333333]
[0.         0.44444444 0.88888889 1.33333333]
[0.         0.44444444 0.88888889 1.33333333]
[0.         0.44444444 0.88888889 1.33333333]
[0.         0.44444444 0.88888889 1.33333333]
```

**Reply by bheiles** [reply]
Hi @MatthewFlamm. I have been experimenting with your examples and tried integrating all that in my workflow. I have discovered something peculiar about the integration time.

If i take your first example


`grid = pyvista.UniformGrid((11, 11, 11), (1, 1, np.sqrt(2)))...`
`velocity = np.zeros((grid.n_points, 3))`

run it with a low `velocity[:,0] = 0.0001`, I fail to enforce the max_time condition I set:

```
stream = grid.streamlines(  
    start_position=(0, 5, 5),  
    max_time=3,  
    vectors="velocity",  
    integrator_type=2,  
    initial_step_length=0.1,   
    integration_direction="forward",   
    step_unit="l"  
)
```
`stream["IntegrationTime"][-5:]`
`array([26000., 27000., 28000., 29000., 30000.])`

If I change the `initial_step_length`

```
stream = grid.streamlines(  
    start_position=(0, 5, 5),  
    max_time=3,  
    vectors="velocity",  
    integrator_type=2,  
    initial_step_length=0.0001,   
    integration_direction="forward",   
    step_unit="l"  
)
```

`stream["IntegrationTime"][-5:]`
`array([1997., 1998., 1999., 2000., 2001.])`


It also happens with higher velocities such as `velocity[:,0]=0.01`:
`stream["IntegrationTime"][-5:]`
`array([19.97, 19.98, 19.99, 20.  , 20.01])`

**Reply by MatthewFlamm** [reply]
`max_time` is also in `step_unit` so your first example has a max integration length of 3 physical units.  So these results make sense to me.

**Reply by bheiles** [reply]
Ok then `max_time` does not have anything to do with time but relates to length. That's what was confusing me... So to somehow figure out the max integration time the algorithm is going to output I should do max_time/velocity right?

**Reply by MatthewFlamm** [reply]
vtk used to support the units in actual time units, but that was removed at some point.

If your `step_unit` is in physical units, then yes.  But any real flow has varying velocities in the domain so it isn't that straightforward.  I would align the maximum integration time, if needed, in seconds afterwards like in the example I gave.

**Reply by bheiles** [reply]
Hmm yeah the problem is that Im trying to generate thousands of streamlines in a complex vasculature so I cant adjust each parameter every time I change my seed point... Thanks for your help though!
